### ISSUE TITLE
Getting File ID from name

### ISSUE THREAD

**Issue by selo2020** [Backblaze/b2-sdk-python#143]
Hi,

I know that this question was asked previously [here](https://github.com/Backblaze/b2-sdk-python/issues/122) but I had some further questions regarding this functionality. 

You previously mentioned that we could use `list_file_versions` to do this - which is fairly easy to do (just limit the fetch_count to 1). But my concern with that is that in terms of "API cost" it's more expensive than calling `get_file_id` - as `list_file_versions` is in the same class as list_file_names. 

So my question is, could you have a call which assumes the latest version for the file you want the information about? 

If this is resolved with some other solution then I would be happy to use that but I am just trying to reduce the number of API calls and I couldn't see any obvious alternative. 

Thanks

**Comment by ppolewicz** [comment]
I get what you are saying, but there is no `get_file_id` in the list of b2 api available operations. If the server would expose such operation, we would expose it as well in the sdk interface, but as of now, there is not much b2-sdk-python can do (unless I misunderstood something?)

**Comment by selo2020** [comment]
Ok that's cool. Is there any recommended way that we can get similar information without the requirement of a `list_file_versions` call? Or is that the only way right now? (or caching from a `list_file_names` call)

**Comment by ppolewicz** [comment]
Before I answer that, could you please describe the entire operation that you are trying to perform? I bet it's not just to delete a file by name but actually something much bigger is going on. If I can understand that, it would let me to help you better.

**Comment by selo2020** [comment]
I am trying to download a file by name but I want to find out some file metadata before doing so - crucially the size and modification date. Although this is only one part of it - I would also like to be able to get the metadata for a specific file (almost like a stat command). I would rather not cache the listings in memory either as there could be hundreds of thousands of files returned. 

I am aware that there are calls to be able to download a file but I am trying to perform this task without utilising local storage (as we cannot guarantee there is a large amount of storage available) or by putting it all in memory (potentially some files are >100GB in size). So I am trying to implement a range downloader to allow me to "consume" the file, but to do that I need to know the size in advance. Generally I know the file exists, but there could also be the instance where it does not.

Bascially, I am thinking of an operation akin to HeadObject in S3 where you can specify the Key and optionally a VersionId if you need to query a specific version.

**Comment by ppolewicz** [comment]
Why are you checking the metadata of a file before downloading it? If the metadata doesn't match a condition, will you select an older version?

Are you downloading/consuming a single file or many files in part of one big operation?

Can there be hundreds of thousands of files with the same name, or different names?

Actually B2 now supports the S3 interface, so maybe you can just HeadObject itself (though if you are processing many files, it will not be optimal)?

**Comment by selo2020** [comment]
I need to know the size of it in advance because I need to be able to tell the other parts of the application how much data to expect when performing operations on it. Some of the applications will need to know this in advance (others might not care). 

In terms of actual operation its up in the air right now as what I am trying to do is map the B2 to our internal location framework. File counts + names could vary quite substantially depending on the available bucket.

I did think that about S3 but this would only be true of new buckets as older buckets are not S3 compatible.

**Comment by ppolewicz** [comment]
You can get up to 1000 file version entries for the same cost (as if you were getting a single entry) and those contain all the metadata you need. If you can batch your operation so that it lists everything first and then performs the download operation (like `sync` does - you can probably reuse a good portion of it and just customize the filter), then it should be less resource intensive.

As for older buckets not being compatible with S3, it is interesting. I need to think about it for a few days.

**Comment by selo2020** [comment]
Having spoken with some people at Backblaze directly they explained that if you call  b2_download_file_by_name with a HEAD request rather than a GET you'll only receive the metadata for the file (mirroring the HeadObject S3 call).

Given that I should be able to implement that with this the lower level APIs in the B2SDK, but it would be useful if this could be made a direct method of the bucket in a future release. 

Thanks!

**Comment by ppolewicz** [comment]
I see. What should be the name for that method (in the Bucket class)?

**Comment by selo2020** [comment]
Well, I guess it could be similar to the existing `get_file_info` (as I guess there is no point arbitrarily limiting the response to just the ID) so it could just be `get_file_info_by_name` - although that is a bit wordy. 

It does have the benefit of maintaining consistency with the `download_file_by_name` call though. But there is probably a better name that could be used.

**Comment by ppolewicz** [comment]
@bwbeach what do you think? `get_file_info_by_name` sounds good to me. HEAD for that url is documented, so we can use it.

https://github.com/Backblaze/b2-sdk-python/blob/6d35959ccfb06c3888702d91f3ce8d699f24f8e0/b2sdk/api.py#L429 indicates that there was some thought put into returning a FileVersion object that the caller could interact with. Also, `b2sdk.v2` is coming.

I think we should consider changing the return type of `B2Api.get_file_info` making it consistent with `Bucket.get_file_info_by_name()` while we implement this. `Bucket.get_file_info_by_name()` could be available in v1, too. The change seems small and easy to test.

@selo2020 since you have dug into this problem quite deeply already, perhaps you would like to contribute the change to b2sdk instead of rolling your own implementation? There are [many advantages](https://b2-sdk-python.readthedocs.io/en/master/#why-use-b2sdk) of keeping it in one place!

**Comment by selo2020** [comment]
Yes, of course, happy to do that. I'll give that a go and will let you know when it's done.

**Comment by certainmagic** [comment]
D'oh!  I never managed to send my email on this topic.  Sorry.
I think you're going in the right direction -- or at least a direction
that's consistent with b2-sdk-java.  :)

fwiw, b2-sdk-java has this method on the B2StorageClient interface[1]:
  B2FileVersion getFileInfoByName(B2GetFileInfoByNameRequest request) throws
B2Exception;

And the implementation is to do a HEAD call as you described above.  The
implementation
is in B2StorageClientWebifierImpl.getFileInfoByName().

ttfn,
ab

[1]
https://github.com/Backblaze/b2-sdk-java/blob/master/core/src/main/java/com/backblaze/b2/client/B2StorageClient.java#L587
[2]
https://github.com/Backblaze/b2-sdk-java/blob/master/core/src/main/java/com/backblaze/b2/client/B2StorageClientWebifierImpl.java#L494

**Comment by ehossack** [comment]
Just thought I'd mention that I've "rolled my own solution" in the meantime while this issue is being worked on.
If there's any updates or you need a hacky starting point, here's a bump :)

**Comment by ppolewicz** [comment]
Implemented by https://github.com/Backblaze/b2-sdk-python/pull/172
